Event, User, Attendance and Invitation Feature

Misori-simon opened the pull request:
### This pull request contains the following changes
- Created an Event Model with a description, a date, and a creator
 - - Added relevant controller methods for the related views
- Created a User Model with a name
 - - Added relevant controller methods for the related views
- Established a one to many relationships between user and events
 - - Added relevant controller methods for the related routes
- Created an Attendance Model as a many to many relations between users and events
 - Added relevant controller methods for the related views
- Created an Invitation Model as a many to many relations between users and events
- Created Tests for the User, Events, Attendance, and invitations
- Added styling for the project UI

Misori-simon committed:
project setup, scaffold user model

Commit:
Add sessions

Commit:
Add user name to session

Misori-simon committed:
Add events model

add events model with creator id as foreign key

Misori-simon committed:
arranged association between user and events

Misori-simon committed:
added is_signed_in predicate method to check if user is signed in

Commit:
Add attendance join table

Commit:
Add Attendance in View

Misori-simon committed:
add scope for events
-add past events scope
-add upcoming events scope

Misori-simon committed:
add style, organised links

Commit:
Update styles

Commit:
Add rspec

Misori-simon committed:
Add restrictions for attendance, generate model invitations
- restricted only the those not attending an event to be able to attend
- add create method and destroy to the invitations model, with the many to many relations

Commit:
Remove Invitations upon Accepting or Declining

Misori-simon committed:
add test for attendance, invitations
-add test for validity of attendance and invitations
-add test for validity in case of missing variables required for creating invitations and attendance
-add test if a user is attempting to attend or be invited more than ones
-add create events link to menu

Commit:
Fix stylelint

Commit:
Update README

teekaytech commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  # POST /events
+  # POST /events.json
+  def create
+    @event = Event.new(event_params)
+    @event.creator_id = session[:user_id]
```
You should use the  `#build` method on the association,
i.e. `@event = current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

Note that `current_user` is a helper method, which you have to declare in your helper module.

teekaytech commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,55 @@
+
+  # GET /events
+  # GET /events.json
+  def index
+    @events = Event.all
+  end
```
You should list both past and upcoming events in the index action, i.e. 

`@past_events = Event.past_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
`@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

You can then use these instance variables in your view.

teekaytech commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  # GET /users/1
+  # GET /users/1.json
+  def show; end
+
```
lists events created by the user, i.e. 
`@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

lists events previously attended by the user using event model scope, i.e. 
`@past_events = current_user.attended_events.past_events`

lists upcoming events previously attended/invited by the user using event model scope, i.e. 
`@upcoming_events = current_user.attended_events.upcoming_events`

teekaytech reviewed (changes requested):
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__Your project is awesome. You've done an amazing job so far :+1: 
However, some changes are required before it can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] Update your readme file
- In the `To install and run` section, some information is missing:
  - After cloning, ask your user to navigate into the project directory
  - Also, some users need to run `yarn install` after running the `bundle install` command.
- [x] There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it.
      **Resources**
- [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
- [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
- [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- [x] Displaying the error messages for each action their corresponding views leads to unnecessary duplication of code. You can easily refactor by moving the error messages to the application layout so that it's available to all other views from one source.

- [x] You can add the `Attend` button to each event's show page. This allows any user to attend such an event by clicking on the button.

- [x] I left a couple of inline comments for you below. Kindly check and make appropriate changes.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

SuyashFowdar commented:
Hello @teekaytech ,

Thank you for reviewing our project.

I have a question concerning the Review No. 4, "_**You can add the Attend button to each event's show page. This allows any user to attend such an event by clicking on the button.**_".

In the Odin Project requirement, part **Event Attendance No. 9**, it tells us to allow only invited users to attend the events, so we added the button only when the user is invited. Should we add the **Attend** button for events even when the user can't attend it?

It would be of great help to us if this could be clarified. :+1:  Thank you very much. :smiley:

teekaytech commented:
> Hello @teekaytech ,
> 
> Thank you for reviewing our project.
> 
> I have a question concerning the Review No. 4, "_**You can add the Attend button to each event's show page. This allows any user to attend such an event by clicking on the button.**_".
> 
> In the Odin Project requirement, part **Event Attendance No. 9**, it tells us to allow only invited users to attend the events, so we added the button only when the user is invited. Should we add the **Attend** button for events even when the user can't attend it?
> 
> It would be of great help to us if this could be clarified. +1 Thank you very much. smiley

Hello @krishnzzz,

Thanks for reaching out. As mentioned in your reference, it is said that adding this feature (i.e. allowing a user to invite other users) attracts extra credit.
Here's the screenshot from Odin:
![2021-01-22(10-59-14)](https://user-images.githubusercontent.com/26856491/105476526-0a771b00-5ca1-11eb-9e22-7821dfd112e2.png)
This is the project statement:
![2021-01-22(11-00-09)](https://user-images.githubusercontent.com/26856491/105476631-2975ad00-5ca1-11eb-94c6-0e7e8922a427.png)

You did well by implementing the invitation feature. I believe the project will be complete if you allow users to attend events by using the `Attend` button, based on the screenshot above. 

Happy coding!

Misori-simon committed:
Remove logic from views, put errors in flash
- created hepler methods for the events and user models
- directed all forms errors for user, session, and events to flash:notice or flash:alert

Commit:
Fix reviews from comments

ElijahTobs reviewed (approved):
# Status: Approved! :tada: 

## Congratulations! @Misori-simon and @krishnzzz  

Good job implementing the change as requested by the previous TSE.

## You can now merge and submit.

## Happy Coding :computer: 

_Feel free to reach out to me on slack_ **`@Elijah Ayandokun`**
  **Elijah O. Ayandokun**
[Github](https://github.com/ElijahTobs) | [LinkedIn](https://www.linkedin.com/in/ayandokunelijah/) | [Twitter](https://twitter.com/ElijahTobs)
